plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -443,6 +462,57 @@ private static <K, V> Map<K, V> getAll(Cache<K, AtomicReference<V>> cache, Itera
         return result.buildOrThrow();
     }
 
+    private static <K, V> Map<K, V> getAll(
+            Cache<K, AtomicReference<V>> cache,
+            Iterable<K> keys,
+            Function<Collection<K>, Map<K, V>> bulkLoader,
+            Predicate<V> isSufficient,
+            BiFunction<V, V, V> merger)
+    {
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        Map<K, AtomicReference<V>> toLoad = new HashMap<>();
+
+        keys.forEach(key -> {
+            // make sure the value holder is retrieved before the new values are loaded
+            // so that in case of invalidation we will not set the stale value
+            AtomicReference<V> currentValueHolder = uncheckedCacheGet(cache, key, AtomicReference::new);
+            V currentValue = currentValueHolder.get();
+            if (currentValue != null && isSufficient.test(currentValue)) {
+                result.put(key, currentValue);
+            }
+            else {
+                toLoad.put(key, currentValueHolder);
+            }
+        });
+
+        if (toLoad.isEmpty()) {
+            return result.buildOrThrow();
+        }
+
+        Map<K, V> newEntries = bulkLoader.apply(toLoad.keySet());
+        newEntries.forEach((key, newValue) -> {
+            AtomicReference<V> valueHolder = toLoad.get(key);
+            result.put(key, merge(merger, key, valueHolder, valueHolder.get(), newValue));
+        });
+
+        return result.buildOrThrow();
+    }
+
+    private static <K, V> V merge(BiFunction<V, V, V> merger, K key, AtomicReference<V> valueHolder, V oldValue, V newValue)
+    {
+        if (newValue == null) {
+            throw new InvalidCacheLoadException("Failed to return a value for " + key);
```

Comment:
wrong place to throw that, as `merge` doesn't know where `newValue` comes from.

let the caller do this validation, and here just do `requireNonNull(newValue, "newValue is null");`
(this also makes `key` arg unnecessary)

thus the function would simplify to

```
private static <K, V> void merge(BiFunction<V, V, V> merger, AtomicReference<V> valueHolder, V newValue)
    {
        requireNonNull(newValue, "newValue is null");
        valueHolder.accumulateAndGet(newValue, (currentValue, nextValue) -> currentValue == null ? nextValue : merger.apply(currentValue, nextValue));
    }
```

but then it's one-liner and deserves inline

Reply:
I switched to the `requireNonNull(newValue, "newValue is null");` but still, leave this method as the old value is required for one of the usages